### DISCUSSION TITLE
Expose two HID Gamepad/Joystick devices with one board

### DISCUSSION THREAD

**Discussion by brickZA** [earlephilhower/arduino-pico#1562] [Q&A] [answered]
I want to use a Waveshare Pi Pico Zero to connect two standard NES controllers to a computer over a single USB port.

I found this this Teensy targeted [NES_to_USB](https://github.com/GadgetReboot/Arduino/tree/master/Teensy/NES_To_USB) sketch and was able to use it with arduino-pico with minor modifications. However, that only works for a single NES controller.

It looks like this  [MHeironimus ArduinoJoystickLibrary](https://github.com/MHeironimus/ArduinoJoystickLibrary/ library supports straight-forward [multiple HID Joystick devices](https://github.com/MHeironimus/ArduinoJoystickLibrary/blob/master/examples/MultipleJoystickTest/MultipleJoystickTest.ino).  Unfortunately it seems that particular Joystick library only supports ATmega32u4 based Arduino boards.

Is there any way I can do the same with arduino-pico? Is it something that should be supported "out of the box" that I'm missing or a feature that could be added?

Otherwise, I'm happy to try and experiment with editing HID descriptors etc,  but from what I've managed to understand from reading the source of arduino-pico and Adafruit TinyUSB the configuration of the HID descriptors is hard-baked into the TinyUSB initialization code that is called by arduino-pico before any of my sketch's code runs, so I don't know how to hook up the custom HID descriptors. An example of how to get started would be much appreciated.

**Comment by earlephilhower** [comment]
The main core only supports 0 or 1 gamepad.  Nothing technically restricting more than one report, it's just not a normal mode of operation and I think it'd be a little complicated to expose the request for 2 gamepads in a way the core would be able to use before the sketch starts.

I believe you can use the Adafruit_TinyUSB library for this because IIRC you control when USB comes up (vs. as part of the core boot in the main repo).  You might want to ask in the Adafruit repo about it.

**Reply by brickZA** [reply]
@earlephilhower Thanks for the quick response. Just to be sure I understand correctly, let me ask some follow-ups :)

Should I choose the "No USB" option under "USB Stack"  and then set up Adafruit TinyUSB myself or can I select "Adafruit TinyUSB" and then somehow hook into the USB setup performed by arduino-pico?

**Reply by earlephilhower** [reply]
Use the Tools->USB Stack->Adafruit.  They have examples you can look at where they set up HID devices.  Might already have a 2-port one, in fact.  You won't be able to use the `Joystick` library, but it's not too hard to make HID reports manually.

**Reply by brickZA** [reply]
Thanks, got it.

So I'll select Tools->USB Stack->Adafruit TinyUSB, and then I can modify one of the examples from the Adafruit repo, e.g. [hid_gamepad.ino](https://github.com/adafruit/Adafruit_TinyUSB_Arduino/blob/master/examples/HID/hid_gamepad/hid_gamepad.ino) to report multiple HID gamepads. And if I need help with that I should ask in the Adafruit repo about it.

Hopefully I'll get round to trying it this week. Will let you know how it turned out :)

**Reply by brickZA** [reply, marked as answer]
I managed to get it working, wrote it up on arduino stack overflow: https://arduino.stackexchange.com/questions/93577/expose-two-hid-joystick-devices-with-single-rp2040-board-waveshare-pi-pico-zero/

**Reply by earlephilhower** [reply]
Nice!  Weird about the Linux quirks though.  I wonder if there's some other way to expose multiple HID reports to avoid the issue.  But you have a workaround, so that's good enough to get folks going!

**Reply by brickZA** [reply]
Yes, it is strange. As far as I can tell (did a quick test) MacOS has the same behaviour but I was not able to quickly find ways to fix that. To my limited knowlege Linux and MacOS are simply not properly respecting the USB HID spec. I assume they have their reasons -- I don't know what they are ;) I don't have a Windows machine to test on but word-on-the-net is that Windows handles this correctly out of the box.
